### PR TITLE
Contributing

### PR DESCRIPTION
**Pull request by ferranlala**
I updated the contributing document. I tried to make it more relevant and friendly to people new to our community.
The idea is to offer an overview of the development process and gradually introduce our rules, guidelines or other more advanced topics. I linked existing wiki pages that explain topics in detail, while this document offers a brief introduction.

There are still some older wiki pages with interesting content that are not linked here, or have duplicated content. I'll consolidate that in the next PR.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,29 +1,81 @@
...
+Thank you for contributing to [Mixxx](https://mixxx.org/)! Your work helps DJs all over the world! We are a global team that works by consensus and we are excited to have you join us.
 
-Table of Contents
-1. [Orientation](#Orientation)
-   1. [Git Repositories](#Git-Repositories)
-2. [Git Workflow](#Git-Workflow)
-   1. [All Contributors](#All-Contributors)
-   2. [Core Team](#Core-Team)
+All Mixxx contributors, including the core team members, are volunteers. This means that Mixxx is made possible by their passion and commitment to the project on their free time. This is why it's so important to treat everyone with respect and understanding, and be patient if things sometimes move a bit slower than you would like. Don't forget to read and follow our [Code Of Conduct](https://github.com/mixxxdj/mixxx/blob/master/CODE_OF_CONDUCT.md), as it provides a framework for everyone to feel accepted, respected and safe in all their interactions within the mixxx community.
```
Comment (Be-ing) [inline CONTRIBUTING.md:4]:
```suggestion
All Mixxx contributors, including the core team members, are volunteers. This means that Mixxx is made possible by their passion and commitment to the project on their free time. This is why it's so important to treat everyone with respect and understanding, and be patient if things sometimes move a bit slower than you would like. Please read and follow our [Code Of Conduct](https://github.com/mixxxdj/mixxx/blob/master/CODE_OF_CONDUCT.md), as it provides a framework for everyone to feel accepted, respected and safe in all their interactions within the Mixxx community.
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,29 +1,81 @@
...
+We use [Zulip chat](https://mixxx.zulipchat.com/) to discuss and coordinate everything related to Mixxx development (and more!).
 
-## Git Workflow ##
-### All Contributors ###
-* Each feature/bug fix should be done on its own Git branch so they can be reviewed and merged independently. Refer to [Using Git](https://mixxx.org/wiki/doku.php/using_git) for how to do this. Please ask for help on [Zulip](https://mixxx.zulipchat.com/) if you have questions about using Git after reading that page.
-* Commits should be as small as they can while still building. The smaller the commit, the easier it is to review. It also makes it easier to revert if it is later identified as the source of a bug. If you have lots of changes that you need to commit, a [GUI Git client](https://git-scm.com/downloads/guis) can be helpful for picking out specific changes for multiple small commits.
+- We encourage you to say hi on the [introduce yourself channel](https://mixxx.zulipchat.com/#narrow/stream/109123-introduce-yourself) and share what you would like to contribute.
```
Comment (Be-ing) [inline CONTRIBUTING.md:21]:
```suggestion
- We encourage you to say hi on the [introduce yourself stream](https://mixxx.zulipchat.com/#narrow/stream/109123-introduce-yourself) and share what you would like to contribute.
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,29 +1,81 @@
...
+- If you need help with your contribution, the Mixxx codebase or anything else related to Mixxx development don't hesitate to ask on the [development help channel](https://mixxx.zulipchat.com/#narrow/stream/247620-development-help).
```
Comment (Be-ing) [inline CONTRIBUTING.md:22]:
```suggestion
- If you need help with your contribution, the Mixxx codebase or anything else related to Mixxx development don't hesitate to ask on the [development help stream](https://mixxx.zulipchat.com/#narrow/stream/247620-development-help).
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,29 +1,81 @@
...
+- The [development channel](https://mixxx.zulipchat.com/#narrow/stream/109171-development) is where day to day discussion about code, new features and releases happens.
```
Comment (Be-ing) [inline CONTRIBUTING.md:23]:
```suggestion
- The [development stream](https://mixxx.zulipchat.com/#narrow/stream/109171-development) is where day to day discussion about code, new features and releases happens.
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,29 +1,81 @@
...
+
+We also have lots of additional information for users and developers on the [Mixxx wiki](https://github.com/mixxxdj/mixxx/wiki), make sure to check it out if there's some piece of information you can't find.
+
+## How can I contribute? ##
+
+There are lots of ways you can contribute to Mixxx. Even if you don't know how to program, there's plenty of things you can do to help:
+
+ - [Report a bug](https://github.com/mixxxdj/mixxx/wiki/Reporting%20bugs)
+ - [Help testing the latest developments](https://github.com/mixxxdj/mixxx/wiki/Testing)
```
Comment (Be-ing) [inline CONTRIBUTING.md:32]:
```suggestion
 - [Test the latest developments](https://github.com/mixxxdj/mixxx/wiki/Testing)
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,29 +1,81 @@
...
+ - [Help translate Mixxx to other languages](https://github.com/mixxxdj/mixxx/wiki/Internationalization)
```
Comment (Be-ing) [inline CONTRIBUTING.md:33]:
```suggestion
 - [Translate Mixxx to other languages](https://github.com/mixxxdj/mixxx/wiki/Internationalization)
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,29 +1,81 @@
...
+ - [Improve the Mixxx Manual](https://github.com/mixxxdj/manual#readme)
```
Comment (Be-ing) [inline CONTRIBUTING.md:34]:
```suggestion
 - [Improve the Mixxx manual](https://github.com/mixxxdj/manual#readme)
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,29 +1,81 @@
...
+ - [Improve the Mixxx website](https://github.com/mixxxdj/website)
+
+If you have some coding skills, or you are willing to learn them you can do even more:
+
+ - [Add native support for new hardware controllers](https://github.com/mixxxdj/mixxx/wiki/Contributing%20mappings)
```
Comment (Be-ing) [inline CONTRIBUTING.md:39]:
```suggestion
 - [Add a mapping for a new controller](https://github.com/mixxxdj/mixxx/wiki/Contributing%20mappings)
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,29 +1,81 @@
...
+### Pick an issue to work on ###
+You might already have in mind that feature you always dreamed of! If that's not the case or you want to start with something easy, you can ask for ideas on [Zulip](https://mixxx.zulipchat.com/#narrow/stream/247620-development-help) or pick an [easy issue](https://bugs.launchpad.net/mixxx/+bugs?field.tag=easy).
+
+### Time to code ###
+Now it's time to write some code!
+We recommend to [set up an IDE](https://github.com/mixxxdj/mixxx/wiki/Developer%20Tools) to assit you in the code writing process.
+
+We use [Git](https://git-scm.com/) for source control. Git is a distributed version control system that allows us to keep track of how our code changes and work together on it. If you are new to Git or source control in general, you can have a look at [Using Git](https://github.com/mixxxdj/mixxx/wiki/Using%20Git) to learn how.
+
+If you have trouble with any of this remember that you can always ask on the [Zulip development help channel](https://mixxx.zulipchat.com/#narrow/stream/247620-development-help).
```
Comment (Be-ing) [inline CONTRIBUTING.md:61]:
```suggestion
If you have trouble with any of this remember that you can always ask on the [Zulip development help stream](https://mixxx.zulipchat.com/#narrow/stream/247620-development-help).
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -1,29 +1,81 @@
...
+
+### Open a Pull Request ###
+Once you finish coding, it's time to show your changes to other Mixxx developers. You do so by [opening a Pull Request](https://github.com/mixxxdj/mixxx/wiki/Using%20Git#open-a-pull-request).
```
Comment (Be-ing) [inline CONTRIBUTING.md:64]:
```suggestion
Once you finish coding, it's time to show your changes to other Mixxx developers. You do so by [opening a pull request](https://github.com/mixxxdj/mixxx/wiki/Using%20Git#open-a-pull-request).
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -1,29 +1,81 @@
...
+
+We use [Git](https://git-scm.com/) for source control. Git is a distributed version control system that allows us to keep track of how our code changes and work together on it. If you are new to Git or source control in general, you can have a look at [Using Git](https://github.com/mixxxdj/mixxx/wiki/Using%20Git) to learn how.
+
+If you have trouble with any of this remember that you can always ask on the [Zulip development help stream](https://mixxx.zulipchat.com/#narrow/stream/247620-development-help).
+
+### Open a Pull Request ###
+Once you finish coding, it's time to show your changes to other Mixxx developers. You do so by [opening a pull request](https://github.com/mixxxdj/mixxx/wiki/Using%20Git#open-a-pull-request).
+
+### Contributor Agreement ###
+Because Mixxx is distributed through the macOS AppStore, in order to distribute your changes with Mixxx, we need a record of your permission to distribute them under Mixxx's license. You'll have to read and sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ).
```
Comment (Be-ing) [inline CONTRIBUTING.md:67]:
```suggestion
In the past Mixxx was distributed in the Mac App Store, which has terms conflicting with Mixxx's license, the GNU GPL (version 2 or later). In case we decide to distribute Mixxx in Apple's App Store again, in order to distribute your changes with Mixxx, we need a record of your permission to distribute them under Mixxx's license. You'll have to read and sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ).
```

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -63,24 +115,18 @@ All of these are automatically built and deployed by our [Jenkins build servers]
...
 * Please help review other people's pull requests. When others review your pull requests, please return the favor. The continued progress of Mixxx depends on all of us working together. Even if you are not familiar with the area of the code being changed in a pull request, you can be helpful by building the branch, verifying that it works as described, and commenting with feedback about the user experience design.
-* If you demonstrate good coding skills, help review pull requests, contribute major features, and show a commitment to Mixxx over time, we may invite you to the core team.
 
-### Core Team ###
+#### Core Team ####
 Mixxx core team members are contributors who have write access to the [upstream mixxxdj repositories](https://github.com/mixxxdj/) on GitHub, access to the Jenkins web interface for the build servers, and access to the private Zulip stream for the core team.
 
-* Enable [two-factor authentication (2FA)](https://help.github.com/en/github/authenticating-to-github/securing-your-account-with-two-factor-authentication-2fa) for your GitHub account.
```
Comment (Be-ing) [inline CONTRIBUTING.md:77]: Why did you remove this information?

Comment (ferranpujolcamins) [reply CONTRIBUTING.md:77]: I move to the linked wiki page. This is not something the majority of users need to worry about.

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -1,29 +1,81 @@
...
+If you have some coding skills, or you are willing to learn them you can do even more:
+
+ - [Add a mapping for a new controller](https://github.com/mixxxdj/mixxx/wiki/Contributing%20mappings)
+ - [Fix bugs or program new features](#Your-first-Mixxx-code-contribution)
+
+We welcome programmers of all backgrounds and skill levels, and our community will do their best to help you with any problems you encounter.
+
+## Your first Mixxx code contribution ##
+
+### Compile Mixxx for the first time ###
```
Comment (daschuer) [inline CONTRIBUTING.md:46]: We need to make sure that contributors create a GitHub account + a Mixxx fork first. Than clone the private repro.
Following the links here doe not emphasis this.
It is the most common issue in contributions from not experienced GitHub users.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -1,29 +1,81 @@
...
+- [Compiling on macOS](https://github.com/mixxxdj/mixxx/wiki/Compiling%20on%20macOS)
+
+### Pick an issue to work on ###
+You might already have in mind that feature you always dreamed of! If that's not the case or you want to start with something easy, you can ask for ideas on [Zulip](https://mixxx.zulipchat.com/#narrow/stream/247620-development-help) or pick an [easy issue](https://bugs.launchpad.net/mixxx/+bugs?field.tag=easy).
+
+### Time to code ###
+Now it's time to write some code!
+We recommend to [set up an IDE](https://github.com/mixxxdj/mixxx/wiki/Developer%20Tools) to assit you in the code writing process.
+
+We use [Git](https://git-scm.com/) for source control. Git is a distributed version control system that allows us to keep track of how our code changes and work together on it. If you are new to Git or source control in general, you can have a look at [Using Git](https://github.com/mixxxdj/mixxx/wiki/Using%20Git) to learn how.
```
Comment (daschuer) [inline CONTRIBUTING.md:59]: This can be moved up an shorten a bit.
The user most likely reads this file on GitHub.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -1,29 +1,81 @@
 # Mixxx Contribution Guidelines #
-Thank you for contributing to [Mixxx](https://mixxx.org/)! Your work helps DJs all over the world! We are global, all volunteer team that works by consensus and we are excited to have you join us. This document specifies technical aspects of our workflow. For social aspects please refer to [CODE_OF_CONDUCT.md](https://github.com/mixxxdj/mixxx/blob/master/CODE_OF_CONDUCT.md) in this repository. We encourage you to introduce yourself on our [Zulip chat](https://mixxx.zulipchat.com/) before starting to contribute code to Mixxx.
+Thank you for contributing to [Mixxx](https://mixxx.org/)! Your work helps DJs all over the world! We are a global team that works by consensus and we are excited to have you join us.
```
[the remark this answers is not in the document]
Comment (ferranpujolcamins) [reply CONTRIBUTING.md:2]: @Be-ing What do you think?

File: CONTRIBUTING.md (hunk 13 above)
Comment (ferranpujolcamins) [reply CONTRIBUTING.md:46]: I updated the linked wiki pages explaining this. 

We can go further and just instruct people to fork, not to clone our repo. What do you think?

**Comment by Holzhaus** [comment]
Thank you for writing this. However, I don't think this should live in our source tree. I can easily go into the wiki so it can be edited indepenent of releases. IMHO `CONTRIBUTING.md` should just be a short list of bullet points that outline the most important points with links to this very verbose guide for people who want to know more.

File: CONTRIBUTING.md (hunk 13 above)
Comment (daschuer) [reply CONTRIBUTING.md:46]: Yes, good idea

**Comment by ferranpujolcamins** [comment]
@Holzhaus If we make this shorter, how would this be different than our current README? IMHO, either we keep CONTRIBUTING with this little outline of how to do your first contribution or we delete it. If we make it shorter we can just put this content on the readme, which needs some love anyway.

File: CONTRIBUTING.md (hunk 13 above)
Comment (ferranpujolcamins) [reply CONTRIBUTING.md:46]: Done

**Comment by Holzhaus** [comment]
The `README.md` is for users. The `CONTRIBUTING.md` is for contributors. The latter should outline stuff for people who want to contribute a patch or report a bug.

Something like:

> Hi, thanks for contributing to Mixxx! Here's a quick overview of the most important things to keep in mind, or read our "First time contributors" guide (link to wiki here) to get a more in-depth introduction.
>
> - Be polite and respectful to each other (link to COC)
> - Work in feature branches, and make sure to base your work on the appropriate base branch. If you started from the wrong branch, please rebase and change the PR target. (link to wiki)
>    | Branch | What should go into this branch | 
>    | ------- | ------ |
>    | `X.Y` (stable release) | Controller mappings and user-noticable bugfixes only
>    | `X.Y` (beta release) |  All kinds of bugfixes, but no new features or string changes
>    | `main` branch | Everything else (new features, string changes, etc.)
> - Use our pre-commit hooks (link to setup guide)
> - Open a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/) to get early feedback.
>
> If your work involves **C++ code changes**:
> - Use `-DWARNINGS_FATAL=ON` and `-DDEBUG_ASSERTIONS_FATAL=ON` pass.
> - Every commit should build, please fix broken commits by rebasing.
> - Run `ctest -j $(nproc) --output-on-failure` before committing to ensure that all tests pass.
> - ...

None of that is relevant to a regular user, just for contributors.

**Comment by daschuer** [comment]
I can confirm, that this file is a wall of text, a drive by contributors might not read to an end.
I don't like such long text in other project where I just want to contribute an already written patch.

On the other hand it is a nice read and contains much valuable info for people not used to contribute to Foss projects.

I can imagine one will disappear, when our CI fails and they have no clue how to fix that. In this case a long contribute guideline is just another barrier.

Can we have both? A short checklist for new contributors experienced with other projects, and a longer optional text for novice contributors in the same file?

We may also consider to dispose some prose without info entirely.

**Comment by Holzhaus** [comment]
Yes, I think this document can be moved into the wiki and make "read our *First time contributors guide*" a link to it. we can also link individual parts of the document from the `CONTRIBUTING.md`, for people that want more information about a particular point.

**Comment by daschuer** [comment]
My idea was more to have just one document, but change the structure, to allow quick access for the typical use cases

**Comment by Holzhaus** [comment]
I mean if you think it's necessary to keep this document in the source tree, I'm okay with it. I just thought it would be easier to change and correct little things if it lives in the wiki.

If we have such a document in the source tree, I suggest to put a bullet list like I posted at the top, and put the indepth guide below it.

**Comment by daschuer** [comment]
Both work for me either

File: CONTRIBUTING.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (Be-ing) [reply CONTRIBUTING.md:2]: :shrug:

**Comment by Be-ing** [comment]
I like the idea to keep this document in the code repository and have a condensed "quick start guide" and then more details below.

File: CONTRIBUTING.md (hunk 12 above)
Comment (Be-ing) [reply CONTRIBUTING.md:77]: I'd prefer to move documentation into the code repository or manual from the wiki.

Comment (xeruf) [reply CONTRIBUTING.md:77]: I agree with this change, CONTRIBUTING should only be an introduction, and that section does not belong into the manual either

**Review by xeruf — CHANGES_REQUESTED** [review]
Please pick this back up!
Also, for better editing and diffs I would recommend following https://sembr.org

In general, as others mentioned, this needs to be trimmed to the essence, but otherwise good ideas.

**Comment by ferranpujolcamins** [comment]
I'm sorry to leave this unfinished, but I don't have the mental energy to do more writing work right now. It's frustrating to write something and then receive lots of feedback from people on the community that means I have to write some more.

This is not a critique to anyone, feedback has been valid and it's nobodies fault, so please don't use this as _hey, look, another example on how in this project we burn people out by requesting lots and lots of bike-shedding changes_, it's just that I think I won't be able to put more energy on this.

Thanks to the reviewers for your time.
